### DISCUSSION TITLE
Reading registers on a device (ili9341)?

### DISCUSSION THREAD

**Discussion by jdefields** [notro/fbtft#600] [General]
I'm curious if the the fbtft_read_spi function actually works?

I've tried writing the CMD value, followed by reading, and I always gets all zeroes.  I've also tried writing my own read function inside the ili9341 driver which does a write and read in the same transaction.  I actually get non-zero data back, but it's garbage.

The purpose of this exercise is to be able to read the 0x45 register (current scanline).

Any assistance/guidance would be appreciated.

**Comment by notro** [comment]
I see that [fbtft_read_spi](https://elixir.bootlin.com/linux/latest/C/ident/fbtft_read_spi) is only used by [fb_ili9320.c](https://elixir.bootlin.com/linux/latest/source/drivers/staging/fbtft/fb_ili9320.c) so it has only been tested with that driver. It doesn't set D/C so it can't work for ili9341.

This is the code used in the DRM drivers for read commands on controllers that use D/C: [mipi_dbi_typec3_command_read](https://elixir.bootlin.com/linux/latest/C/ident/mipi_dbi_typec3_command_read)

Looking at a dump of the read commands on an ili9341 display using a DRM driver, I see that 45h is read back as zero for me as well: https://github.com/notro/tinydrm/wiki/Development#debug
